### ISSUE TITLE
AdguardDNS syntax - wildcard rules

### ISSUE THREAD

**Issue by jakecharlie** [AdguardTeam/AdGuardDNS#550]
Hello, I use the AdguardDNS list on NextDNS and ControlD, but I realized that this list has its own rules and that only AdguardDNS (https://adguard-dns.io/en/public-dns.html) can solve it.

 I went to test blocking two Amazon telemetry domains used on the Echo Show and Fire TV.

You can see in their list here: https://adguardteam.github.io/AdGuardSDNSFilter/Filters/filter.txt

Domain blocked in the AdguardDNS list (Edit: Both domains has an asterisk * , for some reason they sometimes disappear from this post):
||device-metrics-us*.amazon.com^
||device-metrics-*.amazon.com^

You can see that instead of adguard team blocking **device-metrics-us.amazon.com** and **device-metrics-us-2.amazon.com** probably to save space in the list and make it smaller or to cover possible new tracking domains of the same company, you (Adguard team) block some domains in a single line and put an asterisk * on the domain itself.

I went to test it in the linux terminal and realized that this type of approach works when using your DNS (https://adguard-dns.io/en/public-dns.html), but testing Next DNS or ControlDNS with the AdguardDNS list, these domains are not being blocked.

You use this type of approach in 244 domains, such as:
||sock*-goguardian.pusher.com^
||analytics*.clickdimensions.com^
||dualstack.adbert-web-lbs-*.elb.amazonaws.com^

With that, I would like to ask you from the AdguardDNS team that, in the name of transparency and in order to benefit many other users of other services and who like your adguarddns list, **please clarify what are the specific syntax rules that AdguardDNS uses, because this way it would be much easier for NextDNS and ControlD to implement and support 100% of your list. What is the syntax? What are the wildcard rules?**

Thank you very much

**TERMINAL LINUX (USING ADGUARD DNS):**
acer@acer-Aspire-A315-51:~$ nslookup device-metrics-us.amazon.com
Server: 127.0.0.53
Address: 127.0.0.53#53

Non-authoritative answer:
Name: device-metrics-us.amazon.com
Address: 0.0.0.0
Name: device-metrics-us.amazon.com
Address: ::

acer@acer-Aspire-A315-51:~$ nslookup device-metrics-us-2.amazon.com
Server: 127.0.0.53
Address: 127.0.0.53#53

Non-authoritative answer:
Name: device-metrics-us-2.amazon.com
Address: 0.0.0.0
Name: device-metrics-us-2.amazon.com
Address: ::

**TERMINAL LINUX (USING NEXTDNS OR CONTROLD WITH ADGUARD DNS LIST):**
acer@acer-Aspire-A315-51:~$ nslookup device-metrics-us.amazon.com
Server: 127.0.0.53
Address: 127.0.0.53#53

Non-authoritative answer:
Name: device-metrics-us.amazon.com
Address: 52.1.172.87
Name: device-metrics-us.amazon.com
Address: 3.208.140.212
Name: device-metrics-us.amazon.com
Address: 44.197.43.221

acer@acer-Aspire-A315-51:~$ nslookup device-metrics-us-2.amazon.com
Server: 127.0.0.53
Address: 127.0.0.53#53

Non-authoritative answer:
Name: device-metrics-us-2.amazon.com
Address: 54.157.203.98
Name: device-metrics-us-2.amazon.com
Address: 54.205.24.65
Name: device-metrics-us-2.amazon.com
Address: 54.166.21.152

**Comment by jslawler** [comment]
https://adguard-dns.io/kb/general/dns-filtering-syntax/

**Comment by hagezi** [comment]
As @jslawler said, these are normal AdBlock/AdGuard rules whose syntax is documented.

To convert AdBlock/AdGuard rules with wildcards to domains you have to convert the rule to a RegEx and then use it to query other lists, like the CISCO, Tranco Toplist or other domains blocklists, to get the domains that match this pattern, an example:

AdGuard rule `||analytics*.clickdimensions.com^` to RegEx (quick and dirty):

`echo "||analytics*.clickdimensions.com^" | sed 's/\^//' | sed 's/^||/\(\.\|\^\)/' | sed 's/^|/\^/' | sed 's/\./\\./g' | sed 's/\-/\\-/g' | sed 's/\*/\.\*/g' | sed 's/\%/\[0\-9\]\*/g' | sed 's/$/\$/'`

Result: `(\.|^)analytics.*\.clickdimensions\.com$`

Query Toplists with RegEx pattern to get the domains:

`curl -sL https://raw.githubusercontent.com/hagezi/dns-data-collection/main/top/toplist-merged.txt | grep -P '(\.|^)analytics.*\.clickdimensions\.com$'`

Result:
```
analytics-au.clickdimensions.com
analytics-ca.clickdimensions.com
analytics.clickdimensions.com
analytics-eu.clickdimensions.com
```

Cheers,
Gerd

**Comment by jakecharlie** [comment]
Thank you very much. I will pass this information on to both teams. Is there anything else about the syntax for correctly reading the entire AdguardDNS list that they should know?

**Comment by hagezi** [comment]
Yes, there is a lot more to consider if you want to have a 1:1 image, which is quite time-consuming and guaranteed to end in faslch positive doamins. That's why nobody does it. Usually the unique rules for domain lists are extracted. This is how NextDNS does it.

The lists can contain regex, there are "open" regexes like `||analytics.` (blocks `*.analytics.*`), there can be whitelist entries, for example, block `example.com` completely, but allow `*.a.example.com`:

```
||example.com^
@@|a.example.com^
```

For the rules there are modifiers, like` $thirdparty`, `$denyallow`, ...

**Comment by jakecharlie** [comment]
Thank you very much. I opened two threads (one on NextDNS and one on ControlD) reporting on this. I hope they make those fixes soon.

 I also found this link. Maybe helps: https://adguard.com/kb/general/ad-filtering/create-own-filters/
